File: old/Privault.Code/Models/Entities/PermissionEntity.cs
using SQLite;

// ReSharper disable UnusedAutoPropertyAccessor.Global
// ReSharper disable AutoPropertyCanBeMadeGetOnly.Global
// ReSharper disable PropertyCanBeMadeInitOnly.Global

namespace Privault.Core.Models.Entities;

/// <summary>
/// Repräsentiert die Zugriffsberechtigung eines Benutzers für einen spezifischen Tresoreintrag.
/// Diese Entität verwaltet die hybride Verschlüsselung, indem sie den AES-Entry-Key in RSA-verschlüsselter Form speichert.
/// <para>
/// Eine Berechtigung ist eindeutig über die Kombination aus <see cref="EntryId"/> und <see cref="UserId"/>.
/// </para>
/// <para>
/// <b>Sicherheitskonzept:</b>
/// Jede Permission enthält den 32-Byte AES-Key des Eintrags, der mit dem öffentlichen RSA-Schlüssel des jeweiligen 
/// Empfängers verschlüsselt wurde. Nur der Besitzer des zugehörigen privaten RSA-Schlüssels kann den Entry-Key 
/// extrahieren und somit die Daten lesen.
/// </para>
/// </summary>
[Table("permissions")]
public class PermissionEntity
{
    /// <summary>
    /// Die interne ID (Auto-Increment).
    /// Dient als Primärschlüssel für den Object-Relational Mapper (ORM) benötigt.
    /// </summary>
    [PrimaryKey, AutoIncrement]
    [Column("id")]
    public int Id { get; set; }

    /// <summary>
    /// Die interne ID des zugehörigen Eintrags.
    /// </summary>
    [Indexed(Name = "uk_permissions_entry_id_user_id", Order = 1, Unique = true)] // für Detailansicht
    [Indexed(Name = "uk_permissions_user_id_entry_id", Order = 2, Unique = true)] // für Sync-Pull
    [Column("entry_id")]
    public int EntryId { get; set; }

    /// <summary>
    /// Die lokale ID des Benutzers (<see cref="UserEntity"/>), dem dieser Zugriff gewährt wurde.
    /// </summary>
    [Indexed(Name = "uk_permissions_entry_id_user_id", Order = 2, Unique = true)] // für Detailansicht
    [Indexed(Name = "uk_permissions_user_id_entry_id", Order = 1, Unique = true)] // für Sync-Pull
    [Column("user_id")]
    public int UserId { get; set; }
    
    /// <summary>
    /// Der AES-Entry-Key für den Eintrag (32 Bytes), verschlüsselt mit dem öffentlichen RSA-Key des Benutzers.
    /// <para>
    /// Wenn beim Synchronisieren festgestellt wird, dass der RSA-Schlüssel veraltet ist, wird <c>EncryptedKey</c> geleert.
    /// </para>
    /// </summary>
    [Column("encrypted_key")]
    public string EncryptedKey { get; set; } = string.Empty;

    /// <summary>
    /// Definiert die Berechtigungsstufe des Benutzers für diesen Eintrag.
    /// <list type="bullet">
    /// <item><description><b>0:</b> Kein Zugriff</description></item>
    /// <item><description><b>1:</b> Nur Lesen</description></item>
    /// <item><description><b>2:</b> Lesen und Schreiben</description></item>
    /// <item><description><b>3:</b> Vollzugriff/Besitzerrecht (inkl. Löschen und Berechtigungen verwalten)</description></item>
    /// </list>
    /// </summary>
    [Column("access_level")]
    public int AccessLevel { get; set; }
}

File: lib/models/entities/permission_entity.dart
/// Repräsentiert die Zugriffsberechtigung eines Benutzers für einen spezifischen Tresoreintrag.
class PermissionEntity {

    /// Die interne ID (Auto-Increment in der Datenbank).
    /// Nullable für neue Einträge, bevor sie in die Datenbank geschrieben werden.
    final int? id;

    /// Die interne ID des zugehörigen Eintrags.
    final int entryId;

    /// Die lokale ID des Benutzers, dem dieser Zugriff gewährt wurde.
    final int userId;

    /// Der AES-Entry-Key für den Eintrag (32 Bytes), verschlüsselt mit dem öffentlichen RSA-Key des Benutzers.
    ///
    /// Wenn beim Synchronisieren festgestellt wird, dass der RSA-Schlüssel des Benutzers veraltet ist,
    /// wird dieser Wert geleert, da der Schlüssel nicht mehr entschlüsselt werden kann.
    final String encryptedKey;

    /// Definiert die Berechtigungsstufe des Benutzers für diesen Eintrag.
    /// * **0:** Kein Zugriff
    /// * **1:** Nur Lesen
    /// * **2:** Lesen und Schreiben
    /// * **3:** Vollzugriff/Besitzerrecht (inkl. Löschen und Berechtigungen verwalten)
    final int accessLevel;

    /// Konstruktor
    PermissionEntity({
        this.id, 
        required this.entryId, 
        required this.userId, 
        required this.encryptedKey, 
        required this.accessLevel
    });

    /// Konvertiert eine [PermissionEntity] in eine Map (z.B. für SQLite oder JSON).
    Map<String, dynamic> toMap() {
        return {'id': id, 'entry_id': entryId, 'user_id': userId, 'encrypted_key': encryptedKey, 'access_level': accessLevel};
    }

    /// Erstellt ein [PermissionEntity] Objekt aus einer Map.
    factory PermissionEntity.fromMap(Map<String, dynamic> map) {
        return PermissionEntity(
            id: map['id'] as int?,
            entryId: map['entry_id'] as int,
            userId: map['user_id'] as int,
            encryptedKey: map['encrypted_key'] as String,
            accessLevel: map['access_level'] as int,
        );
    }

    /// Erzeugt eine Kopie des Objekts mit modifizierten Eigenschaften.
    PermissionEntity copyWith({int? id, int? entryId, int? userId, String? encryptedKey, int? accessLevel}) {
        return PermissionEntity(
            id: id ?? this.id,
            entryId: entryId ?? this.entryId,
            userId: userId ?? this.userId,
            encryptedKey: encryptedKey ?? this.encryptedKey,
            accessLevel: accessLevel ?? this.accessLevel,
        );
    }
}
